Server socket Issue #3

LordRekishi opened the pull request:
Add Server and ClientHandler classes. Resolves issue #3

HelenaHalldinITHS reviewed (changes requested):
Error occurs during testing. Please fix this!

LordRekishi commented:
> Error occurs during testing. Please fix this!

Yes, tests worked locally with IntelliJ but not with mvn test. Looking at problem with module-info requires fields in Test directory.

HelenaHalldinITHS commented:
I'm not very familiar with this subject, however I read the link martin sent (https://docs.oracle.com/javase/tutorial/networking/sockets/clientServer.html), and when comparing it to you code I couldn't really understand the way you code works. Could you please sent a link or something as a reference for the way you solved this isuue :)

FelixJacobsen commented:
Working with sockets is unfamiliar to me and it would be helpful if you could refer where you have done your research. 

I would like to see more code coverage with your tests. Maybe one where you throw an exception?

FelixJacobsen reviewed (commented):
Working with sockets is unfamiliar to me and it would be helpful if you could refer where you have done your research.

I would like to see more code coverage with your tests. Maybe one where you throw an exception?

LordRekishi commented:
> I'm not very familiar with this subject, however I read the link martin sent (https://docs.oracle.com/javase/tutorial/networking/sockets/clientServer.html), and when comparing it to you code I couldn't really understand the way you code works. Could you please sent a link or something as a reference for the way you solved this isuue :)

The issue right now is the test. Martin was working with me while I cleaned up the code and sent the push. 

But the basic tutorial I used was this one: https://www.baeldung.com/a-guide-to-java-sockets 
And then I incorporated ExecutorService to handle incoming Client calls instead of using serverSocket.accept() in the while loop. On Martin's suggestion.

LordRekishi commented:
Tests should now be working. Ready for review!

LordRekishi commented:
For more info about modular testing please see: https://sormuras.github.io/blog/2018-09-11-testing-in-the-modular-world.html
More about ExecutorService and threads: https://www.baeldung.com/java-executor-service-tutorial

LordRekishi commented:
Please see this test example made for the tutorial version of my code as well, especially focus on the @BeforeClass section (FYI @BeforeClass is junit4 version of @BeforeAll).

Test example and example of production code from tutorial: https://github.com/eugenp/tutorials/blob/master/core-java-modules/core-java-networking/src/test/java/com/baeldung/socket/SocketEchoMultiIntegrationTest.java

HelenaHalldinITHS commented on src/test/java/org/fungover/storm/server/ServerTest.java:
```diff
@@ -0,0 +1,45 @@
+            e.printStackTrace();
+        }
+
+        Executors.newSingleThreadExecutor().submit(() -> new Server(port).start());
```
This line has a compilation error. Not able to create a Server objec.

HelenaHalldinITHS commented:
When trying to run the tests (to check code coverage) I get a compilation error, however it seams like it works fine when i run "mvn test". 

The compilation error that occures is in ServerTest on line 28 when trying to creat a Server object. "Package 'org.fungover.storm.server' is declared in module 'storm', which does not export it to module 'storm'".

LordRekishi commented:
> When trying to run the tests (to check code coverage) I get a compilation error, however it seams like it works fine when i run "mvn test".
> 
> The compilation error that occures is in ServerTest on line 28 when trying to creat a Server object. "Package 'org.fungover.storm.server' is declared in module 'storm', which does not export it to module 'storm'".

It's a bug in IntelliJ according to Martin and myself. Everything actually works, just comes up red in IntelliJ. If it makes trouble for us in the future we can make it an issue then. But for Issue#3 everything works as it should.

kappsegla commented:
It looks like the current implementation works with maven and also in IntelliJ when compiling the code with maven before trying to run code or tests in IntelliJ. The problem is to build with IntelliJs build tool which fails and also the intellisense marks classes red in test code. Not good.

LordRekishi commented:
> It looks like the current implementation works with maven and also in IntelliJ when compiling the code with maven before trying to run code or tests in IntelliJ. The problem is to build with IntelliJs build tool which fails and also the intellisense marks classes red in test code. Not good.
> 
> Found a solution that seems to fix all problems. Removed the module-info.java file from the test folder and instead added the following configuration for the maven-surfire-plugin that runs the tests.
> 
> ```
>    <plugin>
>             <groupId>org.apache.maven.plugins</groupId>
>             <artifactId>maven-surefire-plugin</artifactId>
>             <version>3.0.0-M5</version>
>             <configuration>
>                 <useModulePath>false</useModulePath>
>             </configuration>
>         </plugin>
>         
>         So when running tests it's not using the module system.
> ```

Ok, I will implement this! Thank you!

LordRekishi commented:
With Martin's solution everything should be working as intended

HelenaHalldinITHS reviewed (approved):
Looks good!
